src/types/data/date.js

```diff
@@ -1,80 +1,168 @@
-    case 'datetime':
-      return 'YYYY-MM-DD hh:mm a';
-    case 'date':
-    default:
-      return 'YYYY-MM-DD';
+  renderPicker() {
+    const {options, value} = this.props;
+    let datetime = unixToDatetime(value);
+    if (!datetime.isValid()) datetime = moment();
+    return (
+      <TetheredComponent
+        classes={{element: 'tether-element-top'}}
+        renderElementTo='body'
+        attachment='top left'
+      >
+        <div className='tether-target' />
+        <DatetimePicker
+          className='form-data-datetime-picker'
+          onFocusCapture={this.handleFocus}
+          onBlur={this.handleBlur}
+          moment={datetime}
+          onChange={this.handlePickerChange}
+          onDone={this.handleBlur}
+          type={options.get('dateType')}
+          prevMonthIcon='fa fa-angle-left'
+          nextMonthIcon='fa fa-angle-right'
+          doneIcon='fa fa-check'
+          dateIcon='fa fa-calendar'
+          timeIcon='fa fa-clock-o'
+        />
+      </TetheredComponent>
+    )
+  }
+
+  handleFocus = () => {
+    clearTimeout(this.blurTimeout);
+    this.setState({showPicker: true});
+  }
+
+  handleBlur = () => {
+    // Wait for possible focus event before handling the blur event.
+    this.blurTimeout = setTimeout(() => {
+      this.setState({showPicker: false}, () => {
+        this.saveInput();
+        this.props.onBlur();
+      });
+    });
+  }
+
+  saveInput = () => {
+    let unixTimestamp = null;
+    if (this.state.input) {
+      let datetime = stringToDatetime(this.state.input);
+      if (datetime.isValid()) {
+        unixTimestamp = datetimeToUnix(datetime);
+      }
+    }
+    if (unixTimestamp !== this.props.value) {
+      this.props.onChange(unixTimestamp);
+    }
+  }
+
+  handlePickerChange = (datetime) => {
+    const unixTimestamp = datetimeToUnix(datetime);
+    this.props.onChange(unixTimestamp);
+  }
+
+  handleInputChange = (event) => {
+    this.setState({input: event.target.value});
   }
 }
 
-function valueToDate(value) {
-  const dateValue = moment(value).valueOf() / 1000;
-  if (Number.isFinite(dateValue)) {
-    return dateValue;
+function createInitialState(props) {
+  let input = '';
+  let datetime = unixToDatetime(props.value);
+  if (datetime.isValid()) {
+    const type = props.options.get('dateType');
+    input = datetimeToString(datetime, type);
   }
-  return null;
+  return {
+    showPicker: false,
+    input
+  };
+}
+
+function unixToDatetime(unixTime) {
+  moment(unixTime === null ? null : unixTime * 1000)
+}
+
+function datetimeToUnix(datetime) {
+  return datetime.valueOf() / 1000;
 }
 
-function dateToString(value, options) {
-  const date = formatDateValue(value);
-  if (date === '') {
-    return '';
+function stringToDatetime(str, type) {
+  if (type === 'time' && isNaN(str)) {
+    // Time strings (e.g. '12:30 pm') can't be parsed w/o specifying a format.
+    return moment(str, getDateFormat(type));
+  } else {
+    return moment(str);
```

Comment:
Specifying a non-ISO or non-RFC 2822 date format via `moment` is now deprecated. The string will be parsed via `new Date(string)` for now, but in the future will be considered an invalid date.  I believe we should support this feature since there are many formats that users would want to use that would no longer available.

We should either call `new Date` if the moment object is not created, or call moment again with a list of formats that we (or the user could) supply.
https://momentjs.com/docs/#/parsing/string-formats/